### PR TITLE
Adds residual chart (issue #486)

### PR DESCRIPTION
**Pull request by ds26gte**
This adds the function residual-chart-from-list that takes xs, ys, and residuals and draws a line plot with stick intervals from each y to its corresponding residual.

### CODE REVIEW

**Comment by schanzer** [comment]
@ds26gte this is really exciting! I tried the following program, and the resulting chart isn't quite what I expected.
```
include chart
xs  = [list: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
ys  = [list: 2, 4, 6, 8,10,12,14,16,18, 20]
res = [list: 1, 2, 4,-1,-2,-4, 0, 4,-3,  3]

series = from-list.residual-chart(xs, ys, res)
c = render-chart(series)
c.display()
```
image

Some of those residuals should be on the other side of the line. Am I doing something wrong here?

File: src/web/js/trove/chart-lib.js
Comment (schanzer) [inline src/web/js/trove/chart-lib.js]: This should be 0, since we ONLY want to see the residuals.

**Comment by shriram** [comment]
@schanzer Nothing wrong, I just think you've forgotten about a little thing called gravity.

**Comment by shriram** [comment]
@schanzer Nothing wrong, I just think you've forgotten about a little thing called gravity.

**Comment by schanzer** [comment]
Ah yes! The residuals are the y', not delta-y. Silly me 🤦‍♂️

**Comment by schanzer** [comment]
@jpolitz I'll defer to you on the design of the code, but this raises no red flags for me. 

But in reading through this, I think we need a lower-level Pyret-facing API that includes some configurability so that we can use intervals in other areas as well. From reading [the documentation](https://developers.google.com/chart/interactive/docs/gallery/intervals), it looks like we want interval-chart to take in a list of xs, ys, a _list of lists_ (each of which represents an interval). I propose that it also expose a method `.type()`, which allows for stick, bar, box, and point intervals. I could also see a universe in which it exposes a `.color()` method, which colors the bars/sticks/boxes/points.

This would allow us to replicate most of the interval types in Google's API, though only one-at-a-time. But that's easily addressed by composing multiple interval charts view `render-charts`, so ultimately this would give us quite a bit of power.

The functionality of `residual-chart` could still exist, being built on top of this either as part of the chart library or implemented via a Bootstrap library.

**Comment by ds26gte** [comment]
> Ah yes! The residuals are the y', not delta-y. Silly me 🤦‍♂️

Corrected. _Residual_ does mean what you expected, so I've corrected the 3rd arg to be _Δy_ rather than _y'_. 

(Of course, please check that the direction of the offset is also as you would expect. I really don't know for sure.)

File: src/web/arr/trove/chart.arr
Comment (schanzer) [inline src/web/arr/trove/chart.arr:1056]: For consistency's sake, shouldn't this be `line-width`?

File: src/web/arr/trove/chart.arr
Comment (schanzer) [inline src/web/arr/trove/chart.arr:1067]: `line-width`?

File: src/web/arr/trove/chart.arr
Comment (schanzer) [inline src/web/arr/trove/chart.arr]: What's the plural `colors` for here? I don't think it's used in the JS implementation

File: src/web/arr/trove/chart.arr
Comment (schanzer) [inline src/web/arr/trove/chart.arr]: same question - what are `colors` for here?

File: src/web/arr/trove/chart.arr
Comment (schanzer) [inline src/web/arr/trove/chart.arr:1068]: should this be `axis-data`, in keeping with Pyret conventions?

File: src/web/arr/trove/chart.arr
Comment (schanzer) [inline src/web/arr/trove/chart.arr]: Can we rename this `interval-chart-from list`, in keeping with the data series renaming?

**Review by schanzer — CHANGES_REQUESTED** [review]
@ds26gte left a few comments for you, but for the most part this looks _awesome!_

@jpolitz hoping you can weigh in on this as well.

File: src/web/arr/trove/chart.arr
Comment (ds26gte) [reply src/web/arr/trove/chart.arr:1056]: Unfortunately, the existing series use camelCase and kebab-case for different fields, and I was trying to be consistent with them.

File: src/web/arr/trove/chart.arr
Comment (ds26gte) [reply src/web/arr/trove/chart.arr]: In case we use multiple residuals?

File: src/web/arr/trove/chart.arr
Comment (schanzer) [reply src/web/arr/trove/chart.arr:1056]: makes sense. @jpolitz any interest in Dorai opening another PR in which these get normalized across the charting library?

File: src/web/arr/trove/chart.arr
Comment (schanzer) [reply src/web/arr/trove/chart.arr]: Multiple residuals should be handled via multiple instances of `interval-chart`, composed via `render-charts(...)`

File: src/web/arr/trove/chart.arr
Comment (schanzer) [inline src/web/arr/trove/chart.arr]: @ds26gte shouldn't this error include boxes as a recommended type?

File: src/web/arr/trove/chart.arr
Comment (schanzer) [inline src/web/arr/trove/chart.arr]: This shouldn't be called residuals now that we've renamed the series.

File: src/web/arr/trove/chart.arr
Comment (schanzer) [inline src/web/arr/trove/chart.arr]: We should remove the "residual" text, and update accordingly

File: src/web/arr/trove/chart.arr
Comment (schanzer) [inline src/web/arr/trove/chart.arr]: Rename

File: src/web/arr/trove/chart.arr
Comment (schanzer) [inline src/web/arr/trove/chart.arr]: Rename

File: src/web/arr/trove/chart.arr
Comment (schanzer) [inline src/web/arr/trove/chart.arr]: rename

File: src/web/arr/trove/chart.arr
Comment (schanzer) [inline src/web/arr/trove/chart.arr]: rename

File: src/web/arr/trove/chart.arr
Comment (schanzer) [inline src/web/arr/trove/chart.arr]: rename

File: src/web/arr/trove/chart.arr
Comment (schanzer) [inline src/web/arr/trove/chart.arr]: rename

File: src/web/arr/trove/chart.arr
Comment (schanzer) [inline src/web/arr/trove/chart.arr]: I think you've already renamed this -- now I'm wondering if Git is just showing me an old version of the file?

**Review by schanzer — CHANGES_REQUESTED** [review]
Minor nits - I think this is good to go once the `residual` naming is gone.

File: src/web/arr/trove/chart.arr
Comment (jpolitz) [reply src/web/arr/trove/chart.arr:1056]: No issue with it, though having it separate is preferred.

**Comment by jpolitz** [comment]
Want to make sure I understand the expected behavior of `interval-chart-from-list`. Here it's drawing the lines only below the dot. Also, the really small delta seems to cause the whole point to not be rendered?

image

File: src/web/js/trove/chart-lib.js
Comment (jpolitz) [inline src/web/js/trove/chart-lib.js]: It's likely that the data should not be passed through uninterpreted. Probably it would be correct to call `toFixnum` on all of these if they are all numbers (and matching the columns above). e.g. https://github.com/brownplt/code.pyret.org/blob/08ad7d3e876030e0013f8b889d0d3e9d2d99d4ea/src/web/js/trove/chart-lib.js#L504

Otherwise, roughnums will cause issues.

File: src/web/js/trove/chart-lib.js
Comment (jpolitz) [inline src/web/js/trove/chart-lib.js:808]: Similarly the widths and such, if they could come in as roughnums from Pyret, need to be converted somehow.

File: src/web/arr/trove/chart.arr
Comment (jpolitz) [reply src/web/arr/trove/chart.arr:1056]: (Separate PR I mean, not separate naming conventions :-))

**Comment by jpolitz** [comment]
Great. Thanks for the tests!

**Pull request merged**
